### DISCUSSION TITLE
Annotating a UserList subclass

### DISCUSSION THREAD

**Discussion by JakobDev** [python/typing#1492] [Q&A] [answered]
I'm using [collections.UserList](https://docs.python.org/3.12/library/collections.html#collections.UserList) to create a custom list class. See this example.
```python
import collections


class TestClass(collections.UserList):
    def __init__(self) -> None:
        super().__init__()

    def lower_all(self) -> None:
        for count, s in enumerate(self.data):
            self.data[count] = s.lower()


test = TestClass()

test.append("Hello")
test.append("World")

test.lower_all()

print(test[0])
```
In this example, TestClass should always be a list of strings. How can I annotate that?

**Comment by Daverball** [comment, marked as answer]
It depends on the Python version, if you're still on 3.8 you will need to add some `if TYPE_CHECKING` blocks to deal with the fact that `ABCMeta` (or rather `collections.abc.MutableSequence`, which `collections.UserList` inherits from) is not subscriptable.

Starting with 3.9 it should be as simple as:

```python
import collections

class TestClass(collections.UserList[str]):
    def __init__(self) -> None:
        super().__init__()

    def lower_all(self) -> None:
        for count, s in enumerate(self.data):
            self.data[count] = s.lower()
```

With 3.8 you would need to do something like the following:

```python
import collections
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    BaseList = collections.UserList[str]
else:
    BaseList = collections.UserList

class TestClass(BaseList):
    def __init__(self) -> None:
        super().__init__()

    def lower_all(self) -> None:
        for count, s in enumerate(self.data):
            self.data[count] = s.lower()
```

**Reply by JakobDev** [reply]
Thanks for the answer. It looks like, this is not supported by mypy. If you check this Code with mypy:
```python
import collections


class TestClass(collections.UserList):
    def __init__(self) -> None:
        super().__init__()

    def lower_all(self) -> None:
        for count, s in enumerate(self.data):
            self.data[count] = s.lower()


test = TestClass()

test.append("Hello")
test.append("World")

test.lower_all()

print(test[0])

test2: list[str] = []

test[0].foo()
test2[0].foo()
```
YOu get only the error for the `test2` but not for `test`.

**Reply by Daverball** [reply]
`collections.UserList` just like `list` is generic. If you omit the type it will default to `Any`, so once a type is `Any` anything will work.

You can see that by using `reveal_type(test[0])` which will be `Any`. This is also why you are not getting any errors inside the `lower_all` method, a `reveal_type(self.data[count])` will also give you `Any`.

So if you want to have the same behavior as with a `list[str]` you will need to specify `collections.UserList[str]` just as in my example code above.

If you want mypy to warn you about unbound generics you can use the `--disallow-any-generics` option. Then you will get an error at the line `class TestClass(collections.UserList):` for not specifying the type of the `UserList`.

**Reply by JakobDev** [reply]
Sorry, this was absolutely stupid from me 🤦 . I just forget to add it while testing. Now it works.
